tests/apps/test_operators.py

```diff
@@ -0,0 +1,159 @@
+        y2_torch = y_torch/self.w2_torch + self.b2_torch
+        w1_torch_grad = grad(y2_torch, self.w1_torch, grad_outputs=torch.ones_like(y2_torch))
+
+        cxt = tc.Context()
+        y_tc = self.x_tc/self.w1_tc + self.b1_tc
+        y_2tc = y_tc/self.w2_tc + self.b2_tc
+        cxt.result = form_of(y_2tc).gradients(tc.tensor.Dense.ones(y_2tc.shape))[hex_id(self.w1_tc)]
+        w1_tc_grad = load_np(HOST.post(ENDPOINT, cxt))
+
+        self.assertTrue((abs(w1_tc_grad-[t.numpy() for t in w1_torch_grad]) < 0.0001).all())
+
+    def testPow(self):
+        y_torch = self.x_torch**self.w1_torch + self.b1_torch
+        y2_torch = y_torch**self.w2_torch + self.b2_torch
+        w1_torch_grad = grad(y2_torch, self.w1_torch, grad_outputs=torch.ones_like(y2_torch))
+
+        cxt = tc.Context()
+        y_tc = self.x_tc**self.w1_tc + self.b1_tc
+        y_2tc = y_tc**self.w2_tc + self.b2_tc
+        cxt.result = form_of(y_2tc).gradients(tc.tensor.Dense.ones(y_2tc.shape))[hex_id(self.w1_tc)]
+        w1_tc_grad = load_np(HOST.post(ENDPOINT, cxt))
+
+        self.assertTrue((abs(w1_tc_grad-[t.numpy() for t in w1_torch_grad]) < 0.0001).all())
+
+    def testMul(self):
+        y_torch = self.x_torch*self.w1_torch + self.b1_torch
+        y2_torch = y_torch*self.w2_torch + self.b2_torch
+        w1_torch_grad = grad(y2_torch, self.w1_torch, grad_outputs=torch.ones_like(y2_torch))
+
+        cxt = tc.Context()
+        y_tc = self.x_tc*self.w1_tc + self.b1_tc
+        y_2tc = y_tc*self.w2_tc + self.b2_tc
+        cxt.result = form_of(y_2tc).gradients(tc.tensor.Dense.ones(y_2tc.shape))[hex_id(self.w1_tc)]
+        w1_tc_grad = load_np(HOST.post(ENDPOINT, cxt))
+
+        self.assertTrue((abs(w1_tc_grad-[t.numpy() for t in w1_torch_grad]) < 0.0001).all())
+
+    def testMatMul(self):
+        y_torch = self.x_torch@self.w1_torch + self.b1_torch
+        y2_torch = y_torch@self.w2_torch + self.b2_torch
+        w1_torch_grad = grad(y2_torch, self.w1_torch, grad_outputs=torch.ones_like(y2_torch))
+
+        cxt = tc.Context()
+        y_tc = self.x_tc@self.w1_tc + self.b1_tc
+        y_2tc = y_tc@self.w2_tc + self.b2_tc
+        cxt.result = form_of(y_2tc).gradients(tc.tensor.Dense.ones(y_2tc.shape))[hex_id(self.w1_tc)]
+        w1_tc_grad = load_np(HOST.post(ENDPOINT, cxt))
+
+        self.assertTrue((abs(w1_tc_grad-[t.numpy() for t in w1_torch_grad]) < 0.0001).all())
+
+    def testExp(self):
+        w_torch = self.w1_torch.exp()
+        y_torch = self.x_torch*w_torch
+        w1_torch_grad = grad(y_torch, self.w1_torch, grad_outputs=torch.ones_like(y_torch))
+
+        cxt = tc.Context()
+        w_tc = self.w1_tc.exp()
+        y_tc = self.x_tc*w_tc
+        cxt.result = form_of(y_tc).gradients(tc.tensor.Dense.ones(y_tc.shape))[hex_id(self.w1_tc)]
+        w1_tc_grad = load_np(HOST.post(ENDPOINT, cxt))
+
+        self.assertTrue((abs(w1_tc_grad-[t.numpy() for t in w1_torch_grad]) < 0.0001).all())
+
+    # TODO: make it works (chain rule)
```

Comment:
There were a few times when I was working in the neural_net.py repo that I decided I must be missing the chain rule, and each time it turned out that the problem was something else (e.g. getting a partial derivative wrong) and the chain rule was being computed implicitly by calling `Operator.derivative` recursively. I think you should step through the result of each `Operator` in the graph (e.g. `cxt.result = w_tc`, then `cxt.result = y_tc`, etc) to identify the point of divergence. It would be clearer to construct the `Operator` graph in a separate `Op`, e.g. `@tc.post my_calculation(...):` and check the return values one-by-one. If it seems like you do need to apply the chain rule explicitly, you can try defining it just in the `Exp` and `MatMul` operators, to check if the result matches your expectations.

Reply:
Here's a simple example of calculating the chain rule:
```python
class Exp(tc.math.operator.Unary):
    @staticmethod
    def diff(x, wrt=None):
        return Exp(x)

    def derivative(self, variable):
        if is_operator(self.subject):
            return self.diff(self.subject, variable) * derivative_of(self.subject, variable)

        return self.diff(self.subject, variable)

def is_operator(state):
    return isinstance(form_of(state), tc.math.Operator)
```

You can follow this same pattern to implement `MatMul` and `Add` in the test, e.g. `y_tc = Add(MatMul(self.x_tc, self.w1_tc), self.w1_tc)`, to check if your intuition about the chain rule is correct.

Incidentally, did you mean to use `b1_torch` and `b1_tc` instead of `w1_torch` and `w1_tc`?

Reply:
You can follow up on this in a separate PR so we can merge in the trig operator gradients now